jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConditionalHandler.java

```diff
@@ -0,0 +1,738 @@
+            _doHandle = MethodHandles.guardWithTest(
+                MethodHandles.dropArguments(checkPath, 1, Response.class, Callback.class),
+                _doHandle,
+                doNotHandle);
+        }
+
+        // If we have predicate conditions, guard the doHandle method with checkPredicates
+        if (!_predicates.isEmpty())
+        {
+            MethodHandle checkPredicates = lookup.findVirtual(ConditionalHandler.class, "checkPredicates", methodType(boolean.class, Request.class)).bindTo(this);
+            _doHandle = MethodHandles.guardWithTest(
+                MethodHandles.dropArguments(checkPredicates, 1, Response.class, Callback.class),
+                _doHandle,
+                doNotHandle);
+        }
+
+        super.doStart();
+    }
+
+    @Override
+    public boolean handle(Request request, Response response, Callback callback) throws Exception
+    {
+        try
+        {
+            // Invoke the _doHandle MethodHandle which will include checking any conditions in its guards.
+            return (boolean)_doHandle.invokeExact(request, response, callback);
+        }
+        catch (Exception e)
+        {
+            throw e;
+        }
+        catch (Throwable t)
+        {
+            Response.writeError(request, response, callback, t);
+            return true;
+        }
+    }
+
+    /**
+     * Handle a request that has met the conditions.
+     * The default implementation calls the {@link Handler#handle(Request, Response, Callback)} method
+     * on the {@link #getHandler() next Handler}.
+     * @param request The request to handle
+     * @param response The response to generate
+     * @param callback The callback for completion
+     * @return True if this handler will complete the callback
+     * @throws Exception If there is a problem handling
+     * @see Handler#handle(Request, Response, Callback)
+     */
+    protected boolean doHandle(Request request, Response response, Callback callback) throws Exception
+    {
+        Handler next = getHandler();
+        return next != null && next.handle(request, response, callback);
+    }
+
+    /**
+     * This method is called when the request has not met the conditions and is not to
+     * be handled by this handler.  The default implementation returns {@code false}.
+     * Derived implementations may send an error response or handle the request differently.
+     * @param request The request to handle
+     * @param response The response to generate
+     * @param callback The callback for completion
+     * @return True if this handler will complete the callback
+     * @throws Exception If there is a problem handling
+     * @see Handler#handle(Request, Response, Callback)
+     */
+    protected boolean doNotHandle(Request request, Response response, Callback callback) throws Exception
```

Comment:
Rename to `didNotApply()`?

Reply:
I kine of  like that `doNotHandle` is clearly the opposite of `doHandle`.    We could call them `doNotApply` and `doApply`, but then `doApply` is a long way from `doHandle`, which follows the `doXxx` pattern we have elsewhere